text/0000-named-tracers.md

```diff
@@ -0,0 +1,104 @@
+ * _Named Tracers_ identified via a name (e.g. _"io.opentelemetry.contrib.mongodb"_) and a version (e.g._"semver:1.0.0"_) which is associated with the Tracer and the Spans it produces.
+ * A `TracerFactory` as the only means of creating a Tracer.
+
+Based on such an identifier, a Sampler could be implemented that discards Spans from certain libraries. Also, by providing a custom Exporter, Span data could be sanitized before it gets processed in a back-end system. However, this is beyond the scope of this proposal, which only provides the fundamental mechanisms.
+
+## Explanation
+
+From a user perspective, working with *Named Tracers* and *TracerFactories* is conceptually similar to how e.g. the [Java logging API](https://docs.oracle.com/javase/7/docs/api/java/util/logging/Logger.html#getLogger(java.lang.String)) and logging frameworks like [log4j](https://www.slf4j.org/apidocs/org/slf4j/LoggerFactory.html) work. In analogy to requesting Logger objects through LoggerFactories, a tracing library would create specific 'Tracer' objects through a 'TracerFactory'.
+
+New Tracers can be created by providing the name and version of an instrumentation library. The version (following the convention proposed in https://github.com/open-telemetry/oteps/pull/38) is basically optional but *should* be supplied since only this information enables following scenarios:
+* Only a specific range of versions of a given instrumentation library need to be suppressed, while other versions are allowed (e.g. due to a bug in those specific versions).
+* Go modules allow multiple versions of the same middleware in a single build so those need to be determined at runtime.
+
+Instead of using plain strings as an argument for creating new Tracers a Resource identifying an instrumentation library is used. Such resources must have a _version_ and _name_ labels (there could be semantic convention definitions for those labels). Version values will follow the conventions proposed [here](https://github.com/open-telemetry/oteps/pull/38).
+
+```java
+// Create a tracer for given instrumentation library in a specific version.
+Tracer tracer = OpenTelemetry.getTracerFactory().getTracer("io.opentelemetry.contrib.mongodb", "semver:1.0.0");
+```
+
+This `TracerFactory` replaces the global `Tracer` singleton object as a ubiquitous point to request a Tracer instance.
+
+If no tracer name (null or empty string) is specified, following the suggestions in ["error handling proposal"](open-telemetry/opentelemetry-specification#153), a "smart default" will be applied and a default tracer implementation is returned.
+
+
+## Internal details
+
+By providing a `TracerFactory` and *Named Tracers*, a vendor or OpenTelemetry implementation gains more flexibility in providing Tracers and which attributes they set in the resulting spans that are produced.
+
+The SpanData class is extended with a `getLibraryResource` function that returns the resource associated with the Tracer that created the span.
```

Comment:
Is `LibraryResouce` a part of SpanData that is supposed to be serialized and propagated when the data leaves the local process and thus it is available for intermediary nodes (such as Agents and Collectors) and final destination/backend to inspect or it is ephemeral and only available in the local process to the exporters and samplers?

If the intent is that this is part of serialized span data then I would expect this proposal to define where exactly in the span data this fits. Is it going to be encoded in Span attributes? Is it going to be a separate `libraryresource` field? If it is a new field how this will be encoded / translated for existing trace protocols that do not have the equivalent concept?

Reply:
Since this `LibraryResource` should be handled the same way as all kinds of ressources, questions about serialization and encoding really need to be answered in that part of the specification https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-resources.md.
The specification on resources probably needs attention anyway since it mentions the API a couple of times and there's a proposal about removing those from the API (https://github.com/open-telemetry/opentelemetry-specification/pull/254) currently. Also serialization and encoding are not addressed at all at the moment.

Reply:
I think the question of serialization can be addressed later.